src/frontend/src/optimizer/plan_node/logical_agg.rs

```diff
@@ -300,52 +301,59 @@ impl LogicalAgg {
     }
 }
 
-/// `ExprHandler` extracts agg calls and references to group columns from select list, in
-/// preparation for generating a plan like `LogicalProject - LogicalAgg - LogicalProject`.
-struct ExprHandler {
-    /// `project` contains all the ExprImpl inside GROUP BY clause (e.g. v1 for group by v1)
-    /// followed by those inside aggregates (e.g. v1 + v2 for min(v1 + v2)).
-    pub project: Vec<ExprImpl>,
-    group_key_len: usize,
-    /// When dedup (rewriting AggCall inputs), it is the index into projects.
-    /// When rewriting InputRef outside AggCall, where it is required to refer to a group column,
-    /// this is the index into LogicalAgg::schema.
-    /// This 2 indices happen to be the same because we always put group exprs at the beginning of
-    /// schema, and they are at the beginning of projects.
-    expr_index: HashMap<ExprImpl, usize>,
-    pub agg_calls: Vec<PlanAggCall>,
-    pub error: Option<ErrorCode>,
+/// `LogicalAggBuilder` extracts agg calls and references to group columns from select list and
+/// build the plan like `LogicalAgg - LogicalProject`.
+/// it is constructed by `group_exprs` and collect and rewrite the expression in selection and
+/// having clause.
+struct LogicalAggBuilder {
+    /// the builder of the input Project
+    input_proj_builder: LogicalProjectBuilder,
+    /// the column [0..input_group_key_num) in the project's output is the group key
```

Comment:
This comment may not be true when there are duplicates in `group_exprs`, and can cause a panic:
```
with t(v1) as (values (1), (2), (1)) select v1 from t group by v1, v1;
```

More details in a prior discussion:
https://github.com/singularity-data/risingwave/pull/1492#discussion_r840190675

Reply:
It is because the colprune can not handle the plan with same group key and I have fixed it. Also add some regression test.